### ISSUE TITLE
API: /api/v1/search/project should not return deleted projects

### ISSUE THREAD

**Issue by mulder999** [DependencyTrack/dependency-track#2127]
### Current Behavior:

`/api/v1/search/project` and `/api/v1/search` returns deleted projects

### Steps to Reproduce:

1. Create a new project
2. Upload a SBOM
3. Search for the project using `/api/v1/search/project` and `/api/v1/project` to confirm everything is working
4. Delete project and wait until it is deleted
5. Search using `/api/v1/project` and confirm the project is no longer found. Wait and redo step 5 otherwise
6. Search using `/api/v1/search/project` (or `/api/v1/search`) -> BUG Project is still returned

### Expected Behavior:

`/api/v1/search/project` and `/api/v1/search` should not return deleted projects

### Environment:
  
  - Dependency-Track Version: 4.6.2
  - Distribution: Docker
  - Database Server: PostgreSQL

**Comment by valentijnscholten** [comment]
Just to be sure, did you see https://github.com/DependencyTrack/dependency-track/issues/1605?

**Comment by mulder999** [comment]
@valentijnscholten thanks for pointing #1605, I missed it. Seems the improvements for 4.5 and 4.6 are still not ok then...

**Comment by mulder999** [comment]
After stop, clean of indexes and restart, only `vulnerabilites` folder is populated. Project search is now always empty. Attempted several times without any difference and no corresponding error in the logs.
Unsure if I would call that a success.

**Comment by nscuro** [comment]
Rebuilding indexes can take a while, depending on how big the portfolio is and how beefy your host system.

See also my comment in https://github.com/DependencyTrack/dependency-track/issues/1605#issuecomment-1124182525

> Another issue is that operations on the index are performed asynchronously in a thread pool of size 1. I think this is done to avoid race conditions. However, the same thread is also busy with mirroring the NVD (and in v4.5.0 EPSS as well). This means index operations are delayed until NVD mirroring is completed, which can take a rather long time, depending on your internet connection and CPU/RAM.

That said, the project indexes should eventually be rebuilt after restart. Is it still not happening?

**Comment by mulder999** [comment]
Hi Niklas, 

Thanks for reaching out.

Components, servicecomponents, cpe and vulnerability do eventually rebuild something but not license nor project (on one of my instance).

Since I am using thin provisioned rclone drive in my production setup, following your message, I even restored a DB backup on a plain basic docker test setup to test again with the data of the instance that has not restored its license and project indexing cache.

After 26h of testing there is still nothing on the projet side. I have only 24 projects in DT and my test machine is a 3.6Ghz Xeon with 18GB RAM / 12 cores which has been sitting mostly idle all the time.  I doubt my host system is the culprit.

Here are the messages logged at start:
```log
2022-11-23 19:46:46 2022-11-23 18:46:46,075 INFO [DefaultObjectGenerator] Initializing default object generator
2022-11-23 19:46:46 2022-11-23 18:46:46,087 INFO [DefaultObjectGenerator] Dispatching event to reindex licenses
2022-11-23 19:46:46 2022-11-23 18:46:46,088 INFO [DefaultObjectGenerator] Dispatching event to reindex projects
2022-11-23 19:46:46 2022-11-23 18:46:46,089 INFO [DefaultObjectGenerator] Dispatching event to reindex components
2022-11-23 19:46:46 2022-11-23 18:46:46,089 INFO [DefaultObjectGenerator] Dispatching event to reindex vulnerablesoftware
```

Got 2 error types likely unrelated in the logs:
```log
2022-11-23 20:54:59 2022-11-23 19:54:59,339 ERROR [MavenMetaAnalyzer] Request failure
2022-11-23 20:54:59 org.xml.sax.SAXParseException: DOCTYPE is disallowed when the feature "http://apache.org/xml/features/disallow-doctype-decl" set to true.
```
```log
2022-11-24 01:46:56 2022-11-24 00:46:56,875 ERROR [OssIndexAnalysisTask] An error occurred decrypting the OSS Index API Token. Skipping
2022-11-24 01:46:56 javax.crypto.BadPaddingException: Given final block not properly padded. Such issues can arise if a bad key is used during decryption.
```

Luckily this problem affects only endpoints rarely used. In fact I believe those endpoints should be rewritten to use postgresql full text search instead of Lucene. Then the support for this legacy and broken file indexes could be dropped.

**Comment by mulder999** [comment]
Bug and project re-indexing is solved with 4.7.0. Hence closing.
